src/Common/src/System/Security/Cryptography/Asn1V2.Serializer.cs

```diff
@@ -0,0 +1,1431 @@
+                        {
+                            throw new AsnSerializationConstraintException(
+                                SR.Format(
+                                    SR.Cryptography_AsnSerializer_UtcTimeTwoDigitYearMaxTooSmall,
+                                    fieldInfo.Name,
+                                    fieldInfo.DeclaringType.FullName,
+                                    serializerFieldData.TwoDigitYearMax));
+                        }
+                    }
+                }
+                else if (attr is GeneralizedTimeAttribute genTimeAttr)
+                {
+                    expectedTypes = new[] { typeof(DateTimeOffset) };
+                    serializerFieldData.TagType = UniversalTagNumber.GeneralizedTime;
+                    serializerFieldData.DisallowGeneralizedTimeFractions = genTimeAttr.DisallowFractions;
+                }
+                else
+                {
+                    Debug.Fail($"Unregistered {nameof(AsnTypeAttribute)} kind: {attr.GetType().FullName}");
+                    throw new CryptographicException();
+                }
+
+                if (!serializerFieldData.IsCollection && Array.IndexOf(expectedTypes, unpackedType) < 0)
+                {
+                    throw new AsnSerializationConstraintException(
+                        SR.Format(
+                            SR.Cryptography_AsnSerializer_UnexpectedTypeForAttribute,
+                            fieldInfo.Name,
+                            fieldInfo.DeclaringType.Namespace,
+                            unpackedType.FullName,
+                            string.Join(", ", expectedTypes.Select(t => t.FullName))));
+                }
+            }
+
+            var defaultValueAttr = fieldInfo?.GetCustomAttribute<DefaultValueAttribute>(false);
+            serializerFieldData.DefaultContents = defaultValueAttr?.EncodedBytes;
+
+            if (serializerFieldData.TagType == 0 && !serializerFieldData.IsAny)
+            {
+                if (unpackedType == typeof(bool))
+                {
+                    serializerFieldData.TagType = UniversalTagNumber.Boolean;
+                }
+                else if (unpackedType == typeof(sbyte) ||
+                         unpackedType == typeof(byte) ||
+                         unpackedType == typeof(short) ||
+                         unpackedType == typeof(ushort) ||
+                         unpackedType == typeof(int) ||
+                         unpackedType == typeof(uint) ||
+                         unpackedType == typeof(long) ||
+                         unpackedType == typeof(ulong))
+                {
+                    serializerFieldData.TagType = UniversalTagNumber.Integer;
+                }
+                else if (unpackedType.IsLayoutSequential)
+                {
+                    serializerFieldData.TagType = UniversalTagNumber.Sequence;
+                }
+                else if (unpackedType == typeof(ReadOnlyMemory<byte>) ||
```

Comment:
I originally had `byte[]`, when the reader used `ReadOnlySpan` instead of `ReadOnlyMemory`.  Then, by changing to ReadOnlyMemory I became able to not have to re-allocate when reading from DER-compatible data.

While the serializer could use Memory with no problem the deserializer would have to make a copy for that, which seems non-intuitive.

`byte[]` could be used here, and is obviously allocating, but seems sub-optimal.  We could leave "sub-optimal" as a code review problem, I suppose...

Reply:
I imagine most developers would see the lifetime of the deserialized object as completely independent from the lifetime of the original input stream. For example, consider a `ReadOnlyMemory<byte>` that is backed by a shared network buffer. You deserialize a `MyObject` from it, then after deserialization you pass the object to a thread pool thread for further processing and let the original thread unwind so that it can process another request. This results in the original `ReadOnlyMemory<byte>` being placed back into the array pool, where its data can be overwritten, and the `MyObject` is now corrupt.